core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -331,17 +435,22 @@ class LogManager(val logDirs: Array[File],
           log.truncateTo(truncateOffset)
           if (needToStopCleaner)
             cleaner.maybeTruncateCheckpoint(log.dir.getParentFile, topicPartition, log.activeSegment.baseOffset)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to truncate the log for $topicPartition due to IOException", e)
+            logDirFailureChannel.maybeAddLogFailureEvent(log.dir.getParent)
```

Comment:
Good point. I have updated both the `LogManager.truncateTo()` and `LogManager.truncateFullyAndStartAt()` so that these two method will call `maybeAddLogFailureEvent()` and throw a KafkaStorageException if IOException is caught.

Based on your previous comments, I get that it is a bit ugly in when and where we should catch IOException or KafkaStorageException. It is also not nice to have code that needs to decide whether we should throw the original KafkaStoragException or encapsulate the original IOException into a KafkaStorageException. After thinking through this issue, I made the following changes to make the logic cleaner

- `LogDirFailureChannel` is now passed to the constructor of `Log` and `LogSegment` so that they can enqueue offline log dir when there is IOException.
- The patch guarantees if `maybeAddLogFailureEvent()` is called before a new KafkaStorageException object is instantiated (except for very few scenarios where we know it is not needed). Thus we do NOT need to call `maybeAddLogFailureEvent()` when KafkaStorageException is caught. 
- And we only need to call `maybeAddLogFailureEvent()` when IOException is caught. Then we can either log the error and swallow the exception, or we can throw a new instance of KafkaStorageException so that the outside code can catch it and generate the proper error in the response.